# Implementierung Daten zum Vergleich

In [2]:
#imports
import pandas as pd
import sqlalchemy
from sqlalchemy import text
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from pandarallel import pandarallel
from tqdm import tqdm
stop_words = set(stopwords.words('english') + stopwords.words('german') + stopwords.words('french') + stopwords.words('spanish'))
nltk.download('wordnet')
from sqlalchemy import create_engine, select
import pprint

pandarallel.initialize(progress_bar=False,nb_workers=20)
tqdm.pandas()

import gensim as ge
import matplotlib.pyplot as plt
import seaborn as sns

import pyLDAvis
from pyLDAvis import gensim
from gensim import  models
import gensim.corpora as corpora
import pyLDAvis.gensim_models
from gensim.test.utils import datapath
from gensim.models import CoherenceModel
from sklearn.model_selection import train_test_split
import pickle
import os
import numpy as np

import itertools
from itertools import permutations
from sklearn import metrics

pyLDAvis.enable_notebook()

[nltk_data] Error loading stopwords: <urlopen error [Errno -3]
[nltk_data]     Temporary failure in name resolution>
[nltk_data] Error loading wordnet: <urlopen error [Errno -3] Temporary
[nltk_data]     failure in name resolution>


INFO: Pandarallel will run on 20 workers.
INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.


In [13]:
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import SGDClassifier
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import multilabel_confusion_matrix
from sklearn.metrics import f1_score
from sklearn.metrics import precision_recall_fscore_support
from sklearn import metrics

import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
#get data from db
engine = create_engine('postgresql+psycopg2://postgres:5050@localhost:5432/postgres')
sql_query_class = 'SELECT dbrecordid, class FROM ke_stage.corpus_small'
df_class = pd.read_sql(sql_query_class, engine)

In [4]:
#Remove stopwords
def remove_special_chars(text):
    import re
    return re.sub('(^\{\")|(\"\}$)|(^\{)|(\}$)', '', text)


<>:4: DeprecationWarning: invalid escape sequence '\{'
<>:4: DeprecationWarning: invalid escape sequence '\{'
/tmp/ipykernel_141697/612895507.py:4: DeprecationWarning: invalid escape sequence '\{'
  return re.sub('(^\{\")|(\"\}$)|(^\{)|(\}$)', '', text)


In [5]:
#lemmatization of the words
def lemma(x):
    import nltk
    from nltk.tokenize import word_tokenize
    from nltk.stem import WordNetLemmatizer
    w_tokenizer = nltk.tokenize.WhitespaceTokenizer()
    lemmatizer = nltk.stem.WordNetLemmatizer()
    x = x.lower()
    x = [lemmatizer.lemmatize(w) for w in w_tokenizer.tokenize(x)]
    return x

In [6]:
#preprocess
def gensim_pre(x):
    import gensim
    x = gensim.utils.simple_preprocess(str(x), deacc=True)
    return x

In [7]:
#create dataframe with preprocessed data
dfs=[]
collist= ["abstract","title"]
conn = engine.connect()
result = conn.execution_options(stream_results=True).execute(select(([text("""dbrecordid,
                                                                              abstract,
                                                                              title
                                                                              FROM  ke_stage.corpus_small
                                                                              """)] )))
while chunk:= result.fetchmany(100000): ## only get x rows at a time
  df = pd.DataFrame (chunk)
  for column in df:
    print(column)
    df[column] = df[column].astype(str)
    df[column] = df[column].parallel_apply(remove_special_chars)
    if column in collist:
        print("lemma")
        df[column] = df[column].parallel_apply(lemma)
        print("gensim")
        df[column] = df[column].parallel_apply(gensim_pre)
        df[column] = df[column].progress_apply(lambda x: ','.join([word for word in x if word not in (stop_words)]))
  dfs.append(df)
df_res = pd.concat(dfs, ignore_index=True)


dbrecordid
abstract
lemma
gensim


100%|██████████| 100000/100000 [00:01<00:00, 55852.88it/s]


title
lemma
gensim


100%|██████████| 100000/100000 [00:00<00:00, 385326.68it/s]


dbrecordid
abstract
lemma
gensim


100%|██████████| 100000/100000 [00:01<00:00, 56667.28it/s]


title
lemma
gensim


100%|██████████| 100000/100000 [00:00<00:00, 383541.99it/s]


dbrecordid
abstract
lemma
gensim


100%|██████████| 100000/100000 [00:01<00:00, 64007.66it/s]


title
lemma
gensim


100%|██████████| 100000/100000 [00:00<00:00, 387266.02it/s]


dbrecordid
abstract
lemma
gensim


100%|██████████| 100000/100000 [00:01<00:00, 63061.07it/s]


title
lemma
gensim


100%|██████████| 100000/100000 [00:00<00:00, 390675.51it/s]


dbrecordid
abstract
lemma
gensim


100%|██████████| 100000/100000 [00:01<00:00, 61689.75it/s]


title
lemma
gensim


100%|██████████| 100000/100000 [00:00<00:00, 378681.61it/s]


In [8]:
#merge data with averbis class
result = pd.merge(df_res, df_class, on=['dbrecordid'], how='inner')

In [9]:
#remove rows with class 'Rest'
result = result[result['class'] != 'Rest']

In [10]:
#get combined tokens from every column
result['combined'] = result[result.columns[1:2]].parallel_apply(lambda x: ','.join(x.astype(str)) ,axis=1)
result = result.drop(['title',
              'abstract'],axis =1 )
result = result[result["combined"].str.len() > 3]

In [11]:
result

dbrecordid  \
0       BASE::ftdoajarticles:oai:doaj.org/article:a9d8...   
1                                    AGRICOLAIND605957333   
2       BASE::ftdoajarticles:oai:doaj.org/article:4279...   
3       BASE::ftdoajarticles:oai:doaj.org/article:4366...   
4       BASE::ftdoajarticles:oai:doaj.org/article:4e41...   
...                                                   ...   
499995  BASE::ftdoajarticles:oai:doaj.org/article:209a...   
499996          BASE::ftorgprints:oai:orgprints.org:34376   
499997  BASE::ftdoajarticles:oai:doaj.org/article:bd58...   
499998  BASE::ftdoajarticles:oai:doaj.org/article:21d0...   
499999                                          M31706602   

                       class  \
0       Umweltwissenschaften   
1       Umweltwissenschaften   
2       Umweltwissenschaften   
3       Umweltwissenschaften   
4       Umweltwissenschaften   
...                      ...   
499995  Umweltwissenschaften   
499996  Umweltwissenschaften   
499997  Umweltwissenschaften   
499998  Umweltwissenschaften   
499999  Umweltwissenschaften   

                                                 combined  
0       gasoline,vehicle,elevated,emission,volatile,or...  
1       harwood,spurfowl,pternistis,harwoodi,endemic,b...  
2       presente,articulo,hace,analisis,via,ciencias,h...  
3       despite,australian,dust,critical,role,regional...  
4       thousand,people,visit,riparian,forest,near,res...  
...                                                   ...  
499995  gambling,never,investigated,greenlandic,adoles...  
499996  aim,project,domino,reduce,external,input,inten...  
499997  biogenic,silica,used,describe,compound,hydrate...  
499998                                               none  
499999  confluence,area,river,network,hot,spot,polluta...  

[397077 rows x 3 columns]

In [19]:
x = result['combined']
y = result['class']

In [20]:
#split into train and test data
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.25)

## train model SGD

In [21]:
sgd_model = Pipeline([
                ('vect', CountVectorizer(lowercase=False,stop_words=None,tokenizer=None)),
                ('tfidf', TfidfTransformer(use_idf=True,norm="l2")),
                ('clf', SGDClassifier(loss='hinge', penalty='l2',
                                        alpha=1e-3, random_state=42,
                                        max_iter=5, tol=None))
   ])

In [22]:
sgd_model.fit(x_train, y_train)

Pipeline(steps=[('vect', CountVectorizer(lowercase=False)),
                ('tfidf', TfidfTransformer()),
                ('clf',
                 SGDClassifier(alpha=0.001, max_iter=5, random_state=42,
                               tol=None))])

In [23]:
y_pred = sgd_model.predict(x_test)

In [24]:
print("Confusion Matrix")
matrix = metrics.confusion_matrix(y_test, y_pred)
print(matrix)

Confusion Matrix
[[16761  1943  3934  2303]
 [ 1723 17339  1814  4110]
 [ 2042  2484 19265   672]
 [  983  1270   602 22025]]


In [25]:
f1_matrix = metrics.classification_report(y_test, y_pred, target_names=['Medizin', 'Landwirtschaft','Umweltwissenschaften','Ernährung'], output_dict=True)
f1_matrix = pd.DataFrame(f1_matrix).transpose()
#f1_matrix.to_csv("sgd_score_cvs\F1_matrix_"+str(size)+".csv")

In [26]:
f1_matrix

precision    recall  f1-score       support
Medizin                0.779255  0.672026  0.721679  24941.000000
Landwirtschaft         0.752691  0.693949  0.722127  24986.000000
Umweltwissenschaften   0.752098  0.787516  0.769400  24463.000000
Ernährung             0.756613  0.885249  0.815892  24880.000000
accuracy               0.759444  0.759444  0.759444      0.759444
macro avg              0.760164  0.759685  0.757275  99270.000000
weighted avg           0.760202  0.759444  0.757164  99270.000000

In [27]:
# Accuracy of the model
accuracy = metrics.accuracy_score(y_test, y_pred)
print('SGD Classifier Accuracy of the model: {:.2f}%'.format(accuracy*100))

SGD Classifier Accuracy of the model: 75.94%


## train model LDA

In [12]:
#split into train and test data
df_train, df_test = train_test_split(result, test_size=0.25)

In [13]:
#get list of keywords
keywords_train = to_data(df_train)

100%|██████████| 297807/297807 [00:04<00:00, 73906.77it/s] 


In [14]:
# Create Dictionary
id2word = corpora.Dictionary(keywords_train)
# Create Corpus
keywords_str = keywords_train
# Term Document Frequency
corpus = [id2word.doc2bow(text) for text in keywords_str]
# View
print(corpus[:10])

[[(0, 1), (1, 1), (2, 1), (3, 1), (4, 7), (5, 1), (6, 1), (7, 1), (8, 1), (9, 2), (10, 1), (11, 1), (12, 2), (13, 2), (14, 1), (15, 1), (16, 1), (17, 1), (18, 5), (19, 1), (20, 1), (21, 2), (22, 1), (23, 1), (24, 1), (25, 7), (26, 5), (27, 6), (28, 1), (29, 2), (30, 1), (31, 2), (32, 1), (33, 1), (34, 2), (35, 1), (36, 1), (37, 13), (38, 1), (39, 1), (40, 1), (41, 1), (42, 1), (43, 1), (44, 1), (45, 1), (46, 1), (47, 1), (48, 1), (49, 1), (50, 1), (51, 3), (52, 1), (53, 1), (54, 1), (55, 1), (56, 1), (57, 1), (58, 3), (59, 1), (60, 1), (61, 1), (62, 5), (63, 1), (64, 3)], [(13, 2), (29, 1), (55, 1), (56, 1), (63, 2), (64, 8), (65, 1), (66, 1), (67, 1), (68, 1), (69, 1), (70, 1), (71, 3), (72, 1), (73, 1), (74, 2), (75, 1), (76, 1), (77, 1), (78, 3), (79, 1), (80, 1), (81, 1), (82, 1), (83, 1), (84, 1), (85, 1), (86, 1), (87, 1), (88, 1), (89, 1), (90, 3), (91, 1), (92, 1), (93, 2), (94, 1), (95, 1), (96, 1), (97, 5), (98, 1), (99, 2), (100, 1), (101, 1), (102, 2), (103, 1), (104, 1), (

In [15]:
# number of topics
num_topics = 4
# Build LDA model
lda_model = ge.models.LdaMulticore(
                                    corpus=corpus,
                                    id2word=id2word,
                                    num_topics=num_topics,
                                    )
# Print the keywords in the 5 topics
print(lda_model.print_topics())
doc_lda = lda_model[corpus]

[(0, '0.008*"wa" + 0.006*"study" + 0.004*"cell" + 0.003*"gene" + 0.003*"analysis" + 0.003*"result" + 0.003*"data" + 0.003*"using" + 0.003*"level" + 0.002*"high"'), (1, '0.011*"wa" + 0.009*"cell" + 0.006*"protein" + 0.004*"study" + 0.004*"effect" + 0.003*"activity" + 0.003*"result" + 0.003*"water" + 0.003*"using" + 0.003*"level"'), (2, '0.016*"wa" + 0.006*"study" + 0.003*"effect" + 0.003*"high" + 0.003*"used" + 0.003*"concentration" + 0.003*"result" + 0.003*"two" + 0.003*"specie" + 0.003*"activity"'), (3, '0.017*"wa" + 0.009*"patient" + 0.007*"study" + 0.004*"group" + 0.003*"result" + 0.003*"level" + 0.003*"using" + 0.003*"effect" + 0.003*"analysis" + 0.003*"model"')]


In [16]:
# Compute Coherence Score
coherence_model_lda = CoherenceModel(model=lda_model, texts=keywords_train, dictionary=id2word, coherence='c_v')
coherence_lda = coherence_model_lda.get_coherence()
print('Coherence Score: ', coherence_lda)
print('\nPerplexity: ', lda_model.log_perplexity(corpus)) # a measure of how good the model is. lower the better.

Coherence Score:  0.36468614537058963

Perplexity:  -9.106080092428977


In [17]:
LDAvis_data_filepath = os.path.join('/home/ubuntu/ullrich/my_code/data/ldavis_prepared_keywords_max')

if 1 == 1:
    LDAvis_prepared = pyLDAvis.gensim_models.prepare(lda_model, corpus, id2word)
    with open(LDAvis_data_filepath, 'wb') as f:
        pickle.dump(LDAvis_prepared, f)
# load the pre-prepared pyLDAvis data from disk
with open(LDAvis_data_filepath, 'rb') as f:
    LDAvis_prepared = pickle.load(f)
pyLDAvis.save_html(LDAvis_prepared, LDAvis_data_filepath + 'keywords.html')
LDAvis_prepared

PreparedData(topic_coordinates=              x         y  topics  cluster       Freq
topic                                                
3      0.032806 -0.029325       1        1  36.349879
1     -0.061926 -0.000929       2        1  24.195292
0      0.023465  0.041182       3        1  20.056973
2      0.005654 -0.010928       4        1  19.397856, topic_info=          Term           Freq          Total Category  logprob  loglift
2178      cell  125502.000000  125502.000000  Default  30.0000  30.0000
624    protein   77264.000000   77264.000000  Default  29.0000  29.0000
427    patient  137443.000000  137443.000000  Default  28.0000  28.0000
145      plant   43221.000000   43221.000000  Default  27.0000  27.0000
64          wa  449615.000000  449615.000000  Default  26.0000  26.0000
...        ...            ...            ...      ...      ...      ...
180       time   12823.819590   63326.321173   Topic4  -6.2150   0.0430
186      water   12558.915241   61561.867418   Topic4  -6.2359   0.0504
738       data   13038.203809   73397.457973   Topic4  -6.1985  -0.0880
191   analysis   12987.225164   82442.471272   Topic4  -6.2024  -0.2081
624    protein   12521.672588   77264.569562   Topic4  -6.2389  -0.1798

[501 rows x 6 columns], token_table=        Topic      Freq     Term
term                            
103785      2  0.968804  abfalls
886         1  0.234667     acid
886         2  0.326685     acid
886         3  0.133654     acid
886         4  0.304980     acid
...       ...       ...      ...
181851      3  0.033967      کشت
58544       1  0.048957       که
58544       2  0.850070       که
58544       3  0.057858       که
58544       4  0.042281       که

[1351 rows x 3 columns], R=30, lambda_step=0.01, plot_opts={'xlab': 'PC1', 'ylab': 'PC2'}, topic_order=[4, 2, 1, 3])

## implementation testdata

In [ ]:
#load LDA model
temp_file = datapath('/home/ubuntu/ullrich/my_code/data/LDA_model/lda_model_määx')
lda = models.ldamodel.LdaModel.load(temp_file)

In [18]:
# get testdata into list
keywords_test = to_data(df_test)

100%|██████████| 99270/99270 [00:00<00:00, 102892.26it/s]


In [19]:
def to_id_corpus(data):
    # Create Dictionary
    id2word = corpora.Dictionary(data)
    # Term Document Frequency
    corpus = [id2word.doc2bow(text) for text in data]
    return corpus, id2word

In [20]:
#predict topic 
def get_topic(liste, lda):
    to_pro = []
    corpus, id2w = to_id_corpus(liste)
    topic = lda.get_document_topics(corpus, minimum_probability=0.5, minimum_phi_value=None,
                                   per_word_topics=False)
    for t in topic:
            to_pro.append(t)
    return to_pro

In [21]:
topics = get_topic(keywords_test, lda_model)

In [22]:
#split topics into topic-number and certainty-score
df_test['topic'] = topics
df_test["topic"] = df_test["topic"].astype("str")
df_test["topic"] = df_test["topic"].replace(to_replace=r'[^\d|\.|\,]', value='', regex=True)
df_test["topic"] = df_test["topic"].replace('', np.nan)
df_test[["topic","certainty"]] =  df_test["topic"].apply(lambda x: pd.Series(str(x).split(",")))

In [23]:
df_test

dbrecordid  \
168655                                           M1414052   
53121                                AGRICOLAIND606023245   
92343                                 AGRISUS201600118130   
474098                                          M29131189   
122465                                          M30469533   
...                                                   ...   
283871  BASE::ftdoajarticles:oai:doaj.org/article:a728...   
273709                                          M31216106   
420761  BASE::ftdoajarticles:oai:doaj.org/article:c2df...   
137433                                     PSYNDEX0323638   
444660                   BASE::fticrisat:oai:icrisat:8744   

                       class  \
168655            Ernährung   
53121   Umweltwissenschaften   
92343             Ernährung   
474098  Umweltwissenschaften   
122465            Ernährung   
...                      ...   
283871               Medizin   
273709               Medizin   
420761        Landwirtschaft   
137433            Ernährung   
444660        Landwirtschaft   

                                                 combined topic  certainty  
168655  unlabelled,aim,study,wa,prove,whether,solution...   nan        NaN  
53121   preservation,ionic,specie,within,antarctic,ice...     2  0.7036571  
92343   kinetics,cellulose,hydrolysis,extremely,low,ac...     1  0.9933971  
474098  consumer,grade,red,green,blue,depth,rgb,sensor...     3  0.8293239  
122465  background,nurses,typically,required,address,p...     0  0.6368733  
...                                                   ...   ...        ...  
283871  paragangliomas,sao,tumores,neuroendocrinos,rar...   nan        NaN  
273709  dental,pulp,stem,cell,dpscs,stem,cell,apical,p...     0   0.976392  
420761  abstract,background,crispr,cas,systems,provide...   nan        NaN  
137433  introduction,overconfidence,errors,bias,consis...     0  0.5039389  
444660  main,purpose,alternative,seed,delivery,system,...   nan        NaN  

[99270 rows x 5 columns]

In [24]:
df_test.to_csv('/home/ubuntu/ullrich/my_code/data/predicted_LDA_määäx.csv', sep=',')

# Evaluation LDA

In [25]:
#read csv file from predicted topics
df = pd.read_csv('/home/ubuntu/ullrich/my_code/data/predicted_LDA_määäx.csv', sep=',')

In [26]:
# drop unnessecary columns 
df = df.drop(['Unnamed: 0'], axis=1)
# replace topic name
df = df.replace({'class':'Ernährung'},{'class':'Ernährung'})

In [27]:
df

dbrecordid  \
0                                               M1414052   
1                                   AGRICOLAIND606023245   
2                                    AGRISUS201600118130   
3                                              M29131189   
4                                              M30469533   
...                                                  ...   
99265  BASE::ftdoajarticles:oai:doaj.org/article:a728...   
99266                                          M31216106   
99267  BASE::ftdoajarticles:oai:doaj.org/article:c2df...   
99268                                     PSYNDEX0323638   
99269                   BASE::fticrisat:oai:icrisat:8744   

                      class  \
0                 Ernährung   
1      Umweltwissenschaften   
2                 Ernährung   
3      Umweltwissenschaften   
4                 Ernährung   
...                     ...   
99265               Medizin   
99266               Medizin   
99267        Landwirtschaft   
99268             Ernährung   
99269        Landwirtschaft   

                                                combined  topic  certainty  
0      unlabelled,aim,study,wa,prove,whether,solution...    NaN        NaN  
1      preservation,ionic,specie,within,antarctic,ice...    2.0   0.703657  
2      kinetics,cellulose,hydrolysis,extremely,low,ac...    1.0   0.993397  
3      consumer,grade,red,green,blue,depth,rgb,sensor...    3.0   0.829324  
4      background,nurses,typically,required,address,p...    0.0   0.636873  
...                                                  ...    ...        ...  
99265  paragangliomas,sao,tumores,neuroendocrinos,rar...    NaN        NaN  
99266  dental,pulp,stem,cell,dpscs,stem,cell,apical,p...    0.0   0.976392  
99267  abstract,background,crispr,cas,systems,provide...    NaN        NaN  
99268  introduction,overconfidence,errors,bias,consis...    0.0   0.503939  
99269  main,purpose,alternative,seed,delivery,system,...    NaN        NaN  

[99270 rows x 5 columns]

In [ ]:
#df_test = df[(df['certainty'] is not None) & (df['certainty'] >= 0.6 )]

In [28]:
#define all possible combinations of the classes
classes = ['Medizin', 'Landwirtschaft', 'Umweltwissenschaften', 'Ernährung']
topics = [0.0,1.0,2.0,3.0]

unique_combinations = []
permut = itertools.permutations(classes, len(topics))

for comb in permut:
    zipped = zip(comb, topics)
    unique_combinations.append(list(zipped))
 
# printing unique_combination list
print(unique_combinations)

[[('Medizin', 0.0), ('Landwirtschaft', 1.0), ('Umweltwissenschaften', 2.0), ('Ernährung', 3.0)], [('Medizin', 0.0), ('Landwirtschaft', 1.0), ('Ernährung', 2.0), ('Umweltwissenschaften', 3.0)], [('Medizin', 0.0), ('Umweltwissenschaften', 1.0), ('Landwirtschaft', 2.0), ('Ernährung', 3.0)], [('Medizin', 0.0), ('Umweltwissenschaften', 1.0), ('Ernährung', 2.0), ('Landwirtschaft', 3.0)], [('Medizin', 0.0), ('Ernährung', 1.0), ('Landwirtschaft', 2.0), ('Umweltwissenschaften', 3.0)], [('Medizin', 0.0), ('Ernährung', 1.0), ('Umweltwissenschaften', 2.0), ('Landwirtschaft', 3.0)], [('Landwirtschaft', 0.0), ('Medizin', 1.0), ('Umweltwissenschaften', 2.0), ('Ernährung', 3.0)], [('Landwirtschaft', 0.0), ('Medizin', 1.0), ('Ernährung', 2.0), ('Umweltwissenschaften', 3.0)], [('Landwirtschaft', 0.0), ('Umweltwissenschaften', 1.0), ('Medizin', 2.0), ('Ernährung', 3.0)], [('Landwirtschaft', 0.0), ('Umweltwissenschaften', 1.0), ('Ernährung', 2.0), ('Medizin', 3.0)], [('Landwirtschaft', 0.0), ('Ernährung',

In [29]:
#get lists of pred and true values
pred_test = df['topic'].values.tolist()
true_test = df['class'].values.tolist()

In [30]:
#replace items in predicted list
def replace_items(pred_list, true_list):
    liste = list(pred_list)
    for i in range(len(liste)):
        for tupel in true_list:
            if liste[i] == tupel[1]:
                liste[i] = tupel[0]
    return liste

In [31]:
#create f1-score for every combination
def f1_score(combinations, pred_list, true_list):
    result = []
    for combi in combinations:
        res = {}
        pred = replace_items(pred_list,combi)
        f1 = metrics.f1_score(true_list, pred, average='weighted')
        res.update({'Combi': combi,'f1_score': f1})
        result.append(res)
    return result

In [32]:
#get dictionary for each combination and every score
res_dict = f1_score(unique_combinations, pred_test, true_test)

In [33]:
#get highest value of dictionary
def highest_val(lst, key, key2):
    highest_value = None
    for dict in lst:
        score = dict[key]
        combi = dict[key2]  
        if highest_value is None or score > highest_value:
            highest_value = score
            pred_combi = combi
    return highest_value, pred_combi

In [34]:
f1 = highest_val(res_dict,'f1_score','Combi')

In [35]:
f1

(0.17216774674936902,
 [('Medizin', 0.0),
  ('Umweltwissenschaften', 1.0),
  ('Ernährung', 2.0),
  ('Landwirtschaft', 3.0)])

Coherence Score:  0.36468614537058963

Perplexity:  -9.106080092428977

(0.2674848618667972,
 [('Medizin', 0.0),
  ('Landwirtschaft', 1.0),
  ('Umweltwissenschaften', 2.0),
  ('Ernährung', 3.0)])

Coherence Score:  0.33641594383898077

Perplexity:  -7.416605622733863

In [ ]:
#get list of the class and predicted topic
pred_averbis = df[['class', 'topic']].values.tolist()
pred_averbis_test = df_test[['class', 'topic']].values.tolist()

In [ ]:
def calculate_acc_score(pred_class, combinations):
    result = [] 
    for combi in combinations:
        tmp = {}
        total_count = len(pred_class)
        match_count = 0
        #print(combi)
        for item in pred_class:
            item = tuple(item)
            if item in combi:
                match_count = match_count + 1
        #print(match_count)        
        accuracy = match_count / total_count
        tmp.update({'Combi': combi,'match_count': match_count, 'acc_score': accuracy})
        result.append(tmp)
    return result

In [ ]:
#get dict of results
result = calculate_acc_score(pred_averbis, unique_combinations)
result_test = calculate_acc_score(pred_averbis_test, unique_combinations)

In [ ]:
highest_val(result, 'acc_score', 'Combi')
highest_val(result_test, 'acc_score', 'Combi')

(0.23706789355881622,
 [('Ernährung', 0.0),
  ('Landwirtschaft', 1.0),
  ('Rest', 2.0),
  ('Medizin', 3.0),
  ('Umweltwissenschaften', 4.0)])

diagramm

In [ ]:
def count_class_pop(df):
    counted = pd.DataFrame()
    counted['class'] = df['class'].value_counts()
    counted["population"] = counted['class'].values / len(df)
    counted["pop_perc"] = counted['population'].values * 100
    lowest_c = counted.min()['class']
    return counted , lowest_c

In [ ]:
counted, lowest_c = count_class_pop(final_df)

In [ ]:
count_series =  final_df.groupby(['class','topic'])['class'].count()
count_series = count_series.to_frame()
count_series.columns = ['count']
count_series = count_series.reset_index()
print("Anzahl der Publiktationen pro Klasse:" ,int(lowest_c))
ax = sns.barplot(data=count_series,x="count",  y="class", orient="h", hue="topic")
ax.set(xlabel="Menge an Zugewiesenen Topics pro Klasse",ylabel="Klasse(Averbis) & Topic(LDA)")
ax.text(x=0.5, y=1.1, s='Klassifizierungsgenauigkeit des LDA Models anhand des Testdatensatzes',
        fontsize=13, weight='bold', 
        ha='center', va='bottom', transform=ax.transAxes)
ax.text(x=0.5, y=1.05, s="bei einer Testdatensatzgröße von " + str(len(df_test))+ " Publikationen und "+ str(num_topics)+' "Topics"',
        fontsize=8, alpha=0.75, ha='center', va='bottom', transform=ax.transAxes)

fig = ax.get_figure()
fig.savefig("grafiken\\Klasse_zu_pub_"+str(len(count_series))+"_topics_"+str(num_topics)+"4classes.png",dpi=300, bbox_inches = "tight") 
fig.clf()
count_series.to_csv("lda_score_csv\\topic_population_at_"+str(len(count_series))+"_topics_"+str(num_topics)+"4classes.csv")
num_topics = num_topics + 4

size = len(count_series)/2